quil-rs/src/parser/command.rs

```diff
@@ -120,6 +122,58 @@ pub(crate) fn parse_declare<'a>(input: ParserInput<'a>) -> InternalParserResult<
     ))
 }
 
+/// Parse the contents of a `CALL` instruction.
+///
+/// Note, the `CALL` instruction here is unresolved; it can only be resolved within the
+/// full context of a program from an [`crate::instruction::extern_call::ExternSignatureMap`].
+///
+/// Call instructions are of the form:
+///     `CALL @ms{Identifier} @rep[:min 1]{@group{@ms{Identifier} @alt @ms{Memory Reference} @alt @ms{Complex}}}`
+///
+/// For additional detail, see the ["Call" in the Quil specification](https://github.com/quil-lang/quil/blob/7f532c7cdde9f51eae6abe7408cc868fba9f91f6/specgen/spec/sec-other.s).
+pub(crate) fn parse_call<'a>(input: ParserInput<'a>) -> InternalParserResult<'a, Instruction> {
+    let (input, name) = token!(Identifier(v))(input)?;
+
+    let (input, arguments) = many0(parse_call_argument)(input)?;
+    let call = Call { name, arguments };
+
+    Ok((input, Instruction::Call(call)))
+}
+
+fn parse_call_argument<'a>(
+    input: ParserInput<'a>,
+) -> InternalParserResult<'a, UnresolvedCallArgument> {
+    alt((
+        map(
+            parse_memory_reference_with_brackets,
+            UnresolvedCallArgument::MemoryReference,
+        ),
+        map(token!(Identifier(v)), UnresolvedCallArgument::Identifier),
+        map(parse_immediate_value, UnresolvedCallArgument::Immediate),
+    ))(input)
+}
+
+fn parse_immediate_value(input: ParserInput) -> InternalParserResult<Complex64> {
```

Comment:
All of them, actually. All numbers are represented as `Complex64` within Quil `Expression`s, the other place where immediate values are parsed. I suppose the alternative here would be to parse something like an enum with integer, float, and complex variants. That's without precedent and I'm not 100% convinced it'd be worth it.

Reply:
Thanks for that pointer – I found the existing parsing code now, at https://github.com/erichulburd/quil-rs/blob/393-extern_call/quil-rs/src/parser/expression.rs#L78-L102 .  I'd like to share code between these two implementations, since it's fiddly – that was the comment I was going to write but then I couldn't find the other parser! :-)